File: examples/apps/python/msgpacketizer-py/src/msgpacketizerpy/msgpacketizer/subscriber.py
import logging
from typing import Callable, Dict

import msgpack

from ...packetizerpy import packetizer
from .common import Message, to_hex


class Subscriber:
    def __init__(self):
        self._packetizer = packetizer.Subscriber()
        self._callbacks: Dict[int, Callable[[Message], None]] = {}

    def feed(self, buffer: bytes):
        self._packetizer.feed(buffer)

    def subscribe(self, index: int, callback: Callable[[Message], None]):
        self._callbacks[index] = callback
        self._packetizer.subscribe(index, self._callback)

    def unsubscribe(self, index: int):
        del self._callbacks[index]
        self._packetizer.unsubscribe(index)

    def _callback(self, m: packetizer.Message):
        logging.debug(f"callback input index: {m.index}, data: {to_hex(m.data)}")
        unpacked = msgpack.unpackb(m.data, use_list=True, raw=False)  # type: ignore
        logging.debug(f"unpacked: {unpacked}")
        msg = Message(m.index, unpacked)

        if msg.index in self._callbacks:
            self._callbacks[msg.index](msg)
        else:
            logging.warn(f"callback not found for index {msg.index}")

File: examples/apps/go/msgpacketizer-go/pkg/msgpacketizer-go/subscriber.go
package msgpacketizer

import (
	// TODO:
	// "github.com/hideakitai/packetizer-go"
	"packetizer-go"

	"github.com/sirupsen/logrus"
	"github.com/vmihailenco/msgpack/v5"
)

type Subscriber struct {
	subscriber *packetizer.Subscriber
	callbacks  map[uint8]func(Message)
}

func NewSubscriber() *Subscriber {
	return &Subscriber{
		subscriber: packetizer.NewSubscriber(),
		callbacks:  map[uint8]func(Message){},
	}
}

func (s *Subscriber) Feed(buffer []byte) {
	s.subscriber.Feed(buffer)
}

func (s *Subscriber) Subscribe(index uint8, callback func(Message)) {
	s.callbacks[index] = callback
	s.subscriber.Subscribe(index, func(m packetizer.Message) {
		logrus.Debugf("callback input index=%d, data=%s", m.Index, ToHex(m.Data))
		var unpacked interface{}
		err := msgpack.Unmarshal(m.Data, &unpacked)
		if err != nil {
			logrus.Warnf("msgpack unmarshal error: %s", err)
			return
		}
		logrus.Debugf("unpacked: %+v", unpacked)
		msg := Message{m.Index, unpacked}

		if callback, ok := s.callbacks[msg.Index]; ok {
			callback(msg)
		} else {
			logrus.Warnf("callback not found for index=%d", msg.Index)
		}
	})
}

func (s *Subscriber) Unsubscribe(index uint8) {
	delete(s.callbacks, index)
	s.subscriber.Unsubscribe(index)
}
